File: mobile/android/app/src/main/java/com/mesh/wallet/core/security/AppLockController.kt
package com.mesh.wallet.core.security

import androidx.fragment.app.FragmentActivity
import com.mesh.wallet.core.session.WalletSession
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

class AppLockController(private val session: WalletSession) {
    private val _isUnlocked = MutableStateFlow(!needsLaunchLock())
    val isUnlocked: StateFlow<Boolean> = _isUnlocked.asStateFlow()

    private val _didAttemptLaunchBiometric = MutableStateFlow(false)
    val didAttemptLaunchBiometric: StateFlow<Boolean> = _didAttemptLaunchBiometric.asStateFlow()

    private var didEvaluateLaunchLock = needsLaunchLock()

    val shouldShowLock: Boolean
        get() = session.secureStorage.isPasscodeEnabled() && session.hasActiveWallet && !_isUnlocked.value

    fun unlock() {
        _isUnlocked.value = true
    }

    fun unlockForCurrentSession() {
        didEvaluateLaunchLock = true
        unlock()
    }

    suspend fun attemptLaunchBiometricUnlock(activity: FragmentActivity): Boolean {
        if (!shouldShowLock) return false
        if (!session.secureStorage.isBiometricEnabled()) return false
        if (!MeshBiometricAuth.isAvailable(activity)) return false

        _didAttemptLaunchBiometric.value = true
        val result = MeshBiometricAuth.authenticate(activity, "Unlock Mesh")
        if (result == MeshBiometricAuth.AuthResult.SUCCESS) {
            unlock()
            return true
        }
        return false
    }

    private fun needsLaunchLock(): Boolean =
        session.secureStorage.isPasscodeEnabled() && session.hasActiveWallet
}

File: mobile/ios/Mesh/Core/AppLockController.swift
import Combine
import SwiftUI

/// Passcode / Face ID only on cold app launch — not on background return or wallet switch.
@MainActor
final class AppLockController: ObservableObject {
    @Published private(set) var isUnlocked: Bool
    /// Set after a Face ID / Touch ID prompt was shown during cold launch (avoids double prompt).
    @Published private(set) var didAttemptLaunchBiometric = false

    private var didEvaluateLaunchLock = false

    init() {
        WalletSession.reconcile()
        let needsLaunchLock = MeshPasscodeStore.isEnabled && WalletSession.hasActiveWallet
        isUnlocked = !needsLaunchLock
        didEvaluateLaunchLock = needsLaunchLock
    }

    var shouldShowLock: Bool {
        MeshPasscodeStore.isEnabled && WalletSession.hasActiveWallet && !isUnlocked
    }

    /// Once per process launch (opening the app after it was terminated).
    func prepareLaunchLockIfNeeded() {
        guard !didEvaluateLaunchLock else { return }
        didEvaluateLaunchLock = true

        guard MeshPasscodeStore.isEnabled, WalletSession.hasActiveWallet else {
            isUnlocked = true
            return
        }
        isUnlocked = false
    }

    func unlock() {
        isUnlocked = true
    }

    func unlockForCurrentSession() {
        didEvaluateLaunchLock = true
        unlock()
    }

    /// Face ID / Touch ID during splash — keeps passcode UI hidden until this finishes.
    @discardableResult
    func attemptLaunchBiometricUnlock() async -> Bool {
        guard shouldShowLock else { return false }
        guard MeshPasscodeStore.isBiometricEnabled, MeshBiometricAuth.isAvailable else {
            return false
        }

        didAttemptLaunchBiometric = true
        let result = await MeshBiometricAuth.authenticate(
            reason: "Unlock Mesh"
        )
        if result == .success {
            unlock()
            return true
        }
        return false
    }
}
